File: engine/helengine.core/components/AnchorComponent.cs
namespace helengine {
    /// <summary>
    /// Keeps an entity pinned to a parent layout bounds provider and updates its local position when that layout changes.
    /// </summary>
    public class AnchorComponent : Component {
        /// <summary>
        /// Bit flag used to mark a left-edge anchor.
        /// </summary>
        public const byte LeftAnchorFlag = 1;

        /// <summary>
        /// Bit flag used to mark a right-edge anchor.
        /// </summary>
        public const byte RightAnchorFlag = 2;

        /// <summary>
        /// Bit flag used to mark a top-edge anchor.
        /// </summary>
        public const byte TopAnchorFlag = 4;

        /// <summary>
        /// Bit flag used to mark a bottom-edge anchor.
        /// </summary>
        public const byte BottomAnchorFlag = 8;

        /// <summary>
        /// Stores the active anchor edge flags.
        /// </summary>
        public byte AnchorFlags { get; set; }

        /// <summary>
        /// Stores the active anchor distances in left, right, top, bottom order.
        /// </summary>
        public float4 AnchorDistances { get; set; }

        /// <summary>
        /// Tracks the current ancestor bounds provider used to resolve parent-relative anchors.
        /// </summary>
        IAnchorBoundsProvider anchorBoundsProvider;

        /// <summary>
        /// Tracks whether the component is currently subscribed to the window resize fallback.
        /// </summary>
        bool IsSubscribedToWindowResize;

        /// <summary>
        /// Gets a value indicating whether anchoring is currently enabled.
        /// </summary>
        public bool IsAnchored => AnchorFlags != 0;

        /// <summary>
        /// Enables anchoring with specific sides. The entity will maintain its distance from the resolved parent bounds.
        /// </summary>
        /// <param name="left">Anchor to the left edge of the parent bounds.</param>
        /// <param name="right">Anchor to the right edge of the parent bounds.</param>
        /// <param name="top">Anchor to the top edge of the parent bounds.</param>
        /// <param name="bottom">Anchor to the bottom edge of the parent bounds.</param>
        public void EnableAnchoring(bool left = false, bool right = false, bool top = false, bool bottom = false) {
            if (!left && !right && !top && !bottom) {
                DisableAnchoring();
                return;
            }

            if (Parent == null) {
                throw new InvalidOperationException("AnchorComponent must be attached before anchoring can be enabled.");
            }

            RefreshSubscriptions();
            int2 anchorBounds = GetAnchorBounds();
            int2 anchoredSize = GetAnchorSize();
            float3 localPosition = Parent.LocalPosition;

            byte anchorFlags = 0;
            float4 anchorDistances = new float4(0f, 0f, 0f, 0f);

            if (left) {
                anchorFlags |= LeftAnchorFlag;
                anchorDistances.X = localPosition.X;
            }
            if (right) {
                anchorFlags |= RightAnchorFlag;
                anchorDistances.Y = anchorBounds.X - localPosition.X - anchoredSize.X;
            }
            if (top) {
                anchorFlags |= TopAnchorFlag;
                anchorDistances.Z = localPosition.Y;
            }
            if (bottom) {
                anchorFlags |= BottomAnchorFlag;
                anchorDistances.W = anchorBounds.Y - localPosition.Y - anchoredSize.Y;
            }

            AnchorFlags = anchorFlags;
            AnchorDistances = anchorDistances;

            if (Parent != null && Core.Instance != null && Core.Instance.RenderManager3D != null) {
                RefreshSubscriptions();
                RefreshAnchoring();
            }
        }

        /// <summary>
        /// Disables anchoring and stops responding to layout changes.
        /// </summary>
        public void DisableAnchoring() {
            DetachFromBoundsProvider();
            DetachFromWindowResize();
            AnchorFlags = 0;
            AnchorDistances = new float4(0f, 0f, 0f, 0f);
        }

        /// <summary>
        /// Sets anchor distances manually using the resolved parent bounds as the reference frame.
        /// </summary>
        /// <param name="left">Distance from the left edge of the parent bounds.</param>
        /// <param name="right">Distance from the right edge of the parent bounds.</param>
        /// <param name="top">Distance from the top edge of the parent bounds.</param>
        /// <param name="bottom">Distance from the bottom edge of the parent bounds.</param>
        public void SetAnchorDistances(Nullable<float> left = null, Nullable<float> right = null, Nullable<float> top = null, Nullable<float> bottom = null) {
            if (!left.HasValue && !right.HasValue && !top.HasValue && !bottom.HasValue) {
                DisableAnchoring();
                return;
            }

            byte anchorFlags = 0;
            float4 anchorDistances = new float4(0f, 0f, 0f, 0f);
            if (left.HasValue) {
                anchorFlags |= LeftAnchorFlag;
                anchorDistances.X = left.Value;
            }
            if (right.HasValue) {
                anchorFlags |= RightAnchorFlag;
                anchorDistances.Y = right.Value;
            }
            if (top.HasValue) {
                anchorFlags |= TopAnchorFlag;
                anchorDistances.Z = top.Value;
            }
            if (bottom.HasValue) {
                anchorFlags |= BottomAnchorFlag;
                anchorDistances.W = bottom.Value;
            }

            AnchorFlags = anchorFlags;
            AnchorDistances = anchorDistances;

            if (Parent != null && Core.Instance != null && Core.Instance.RenderManager3D != null) {
                RefreshSubscriptions();
                RefreshAnchoring();
            }
        }

        /// <summary>
        /// Refreshes the anchored position from the current bounds provider and stored distances.
        /// </summary>
        public void RefreshAnchoring() {
            if (!IsAnchored || Parent == null) {
                return;
            }

            RefreshSubscriptions();

            int2 anchorBounds = GetAnchorBounds();
            int2 anchorSize = GetAnchorSize();
            float3 localPosition = Parent.LocalPosition;

            if ((AnchorFlags & LeftAnchorFlag) != 0) {
                localPosition.X = AnchorDistances.X;
            } else if ((AnchorFlags & RightAnchorFlag) != 0) {
                localPosition.X = anchorBounds.X - AnchorDistances.Y - anchorSize.X;
            }

            if ((AnchorFlags & TopAnchorFlag) != 0) {
                localPosition.Y = AnchorDistances.Z;
            } else if ((AnchorFlags & BottomAnchorFlag) != 0) {
                localPosition.Y = anchorBounds.Y - AnchorDistances.W - anchorSize.Y;
            }

            Parent.LocalPosition = localPosition;
        }

        /// <summary>
        /// Rebinds to the nearest parent bounds provider when the component is attached to an entity.
        /// </summary>
        /// <param name="entity">Entity receiving the component.</param>
        public override void ComponentAdded(Entity entity) {
            base.ComponentAdded(entity);
            if (IsAnchored) {
                RefreshSubscriptions();
                RefreshAnchoring();
            }
        }

        /// <summary>
        /// Cleans up anchoring resources when the component is removed from its parent entity.
        /// </summary>
        /// <param name="entity">Entity the component was attached to.</param>
        public override void ComponentRemoved(Entity entity) {
            base.ComponentRemoved(entity);
            DisableAnchoring();
        }

        /// <summary>
        /// Rebinds the current provider subscriptions when the parent entity changes enabled state.
        /// </summary>
        /// <param name="newEnabled">New enabled state.</param>
        public override void ParentEnabledChange(bool newEnabled) {
            base.ParentEnabledChange(newEnabled);

            if (!IsAnchored) {
                return;
            }

            if (newEnabled) {
                RefreshSubscriptions();
                RefreshAnchoring();
            } else {
                DetachFromBoundsProvider();
                DetachFromWindowResize();
            }
        }

        /// <summary>
        /// Handles parent bounds changes to reposition anchored entities.
        /// </summary>
        void HandleAnchorBoundsChanged() {
            RefreshAnchoring();
        }

        /// <summary>
        /// Handles window resize fallback events to reposition anchored entities when no parent provider exists.
        /// </summary>
        /// <param name="handle">Window handle reported by the render manager.</param>
        /// <param name="newWidth">Updated host width.</param>
        /// <param name="newHeight">Updated host height.</param>
        void HandleWindowResized(IntPtr handle, int newWidth, int newHeight) {
            RefreshAnchoring();
        }

        /// <summary>
        /// Rebinds the current provider subscriptions to the nearest ancestor layout bounds provider.
        /// </summary>
        void RefreshSubscriptions() {
            IAnchorBoundsProvider newProvider = ResolveAnchorBoundsProvider();

            if (!ReferenceEquals(anchorBoundsProvider, newProvider)) {
                DetachFromBoundsProvider();
                anchorBoundsProvider = newProvider;

                if (anchorBoundsProvider != null) {
                    anchorBoundsProvider.AnchorBoundsChanged += HandleAnchorBoundsChanged;
                }
            }

            if (anchorBoundsProvider == null) {
                AttachToWindowResize();
            } else {
                DetachFromWindowResize();
            }
        }

        /// <summary>
        /// Disconnects the current bounds provider subscription.
        /// </summary>
        void DetachFromBoundsProvider() {
            if (anchorBoundsProvider != null) {
                anchorBoundsProvider.AnchorBoundsChanged -= HandleAnchorBoundsChanged;
                anchorBoundsProvider = null;
            }
        }

        /// <summary>
        /// Connects the fallback window resize subscription if it is not already active.
        /// </summary>
        void AttachToWindowResize() {
            if (IsSubscribedToWindowResize) {
                return;
            }
            if (Core.Instance == null || Core.Instance.RenderManager3D == null) {
                return;
            }

            Core.Instance.RenderManager3D.WindowResized += HandleWindowResized;
            IsSubscribedToWindowResize = true;
        }

        /// <summary>
        /// Disconnects the fallback window resize subscription if it is active.
        /// </summary>
        void DetachFromWindowResize() {
            if (!IsSubscribedToWindowResize) {
                return;
            }

            Core.Instance.RenderManager3D.WindowResized -= HandleWindowResized;
            IsSubscribedToWindowResize = false;
        }

        /// <summary>
        /// Resolves the nearest ancestor bounds provider in the entity chain.
        /// </summary>
        /// <returns>Nearest layout bounds provider when one exists; otherwise null.</returns>
        IAnchorBoundsProvider ResolveAnchorBoundsProvider() {
            Entity current = Parent;

            while (current != null) {
                if (current is IAnchorBoundsProvider provider) {
                    return provider;
                }

                if (current.Components != null) {
                    for (int i = 0; i < current.Components.Count; i++) {
                        if (current.Components[i] is IAnchorBoundsProvider componentProvider) {
                            return componentProvider;
                        }
                    }
                }

                current = current.Parent;
            }

            return null;
        }

        /// <summary>
        /// Resolves the bounds that should be treated as the active anchor space.
        /// </summary>
        /// <returns>Anchor bounds in local pixels.</returns>
        int2 GetAnchorBounds() {
            if (anchorBoundsProvider != null) {
                return anchorBoundsProvider.AnchorBounds;
            }

            return Core.Instance.RenderManager3D.MainWindowSize;
        }

        /// <summary>
        /// Resolves the current size of the anchored entity when it exposes one.
        /// </summary>
        /// <returns>Anchored size in local pixels, or zero when no size provider is available.</returns>
        int2 GetAnchorSize() {
            if (Parent == null) {
                return new int2(0, 0);
            }

            IAnchorSizeProvider bestProvider = null;
            int bestArea = -1;

            if (Parent is IAnchorSizeProvider parentProvider) {
                bestProvider = parentProvider;
                bestArea = GetAnchorArea(parentProvider.AnchorSize);
            }

            for (int i = 0; i < Parent.Components.Count; i++) {
                if (Parent.Components[i] is IAnchorSizeProvider sizeProvider) {
                    int area = GetAnchorArea(sizeProvider.AnchorSize);
                    if (area > bestArea) {
                        bestProvider = sizeProvider;
                        bestArea = area;
                    }
                }
            }

            if (bestProvider == null) {
                return new int2(0, 0);
            }

            return bestProvider.AnchorSize;
        }

        /// <summary>
        /// Computes the area of one size provider result for comparison purposes.
        /// </summary>
        /// <param name="size">Size used to compare providers.</param>
        /// <returns>Signed area used to select the most specific provider.</returns>
        int GetAnchorArea(int2 size) {
            if (size.X < 0 || size.Y < 0) {
                return -1;
            }

            return size.X * size.Y;
        }

        /// <summary>
        /// Gets current anchor configuration as a readable string.
        /// </summary>
        /// <returns>Readable description of the active anchor distances.</returns>
        public string GetAnchorInfo() {
            if (!IsAnchored) {
                return "Not anchored";
            }

            var info = "Anchored to: ";
            var anchors = new List<string>();

            if ((AnchorFlags & LeftAnchorFlag) != 0) anchors.Add($"Left ({AnchorDistances.X:F1}px)");
            if ((AnchorFlags & RightAnchorFlag) != 0) anchors.Add($"Right ({AnchorDistances.Y:F1}px)");
            if ((AnchorFlags & TopAnchorFlag) != 0) anchors.Add($"Top ({AnchorDistances.Z:F1}px)");
            if ((AnchorFlags & BottomAnchorFlag) != 0) anchors.Add($"Bottom ({AnchorDistances.W:F1}px)");

            return info + string.Join(", ", anchors);
        }
    }
}

File: tmp/helengine-core-cpp-regenerated/AnchorComponent.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "AnchorComponent.hpp"
#include "runtime/native_exceptions.hpp"
#include "int2.hpp"
#include "float3.hpp"
#include "Entity.hpp"
#include "AnchorData.hpp"
#include "runtime/native_nullable.hpp"
#include "Component.hpp"
#include "IAnchorBoundsProvider.hpp"
#include "IAnchorSizeProvider.hpp"
#include "runtime/native_string.hpp"
#include "runtime/native_list.hpp"
#include "Core.hpp"
#include "runtime/array.hpp"
#include "runtime/finally.hpp"
#include "runtime/native_cast.hpp"
#include "runtime/native_datetime.hpp"
#include "runtime/native_dictionary.hpp"
#include "runtime/native_disposable.hpp"
#include "runtime/native_enum.hpp"
#include "runtime/native_event.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_list.hpp"
#include "runtime/native_nullable.hpp"
#include "runtime/native_span.hpp"
#include "runtime/native_stack.hpp"
#include "runtime/native_string.hpp"
#include "runtime/native_tuple.hpp"
#include "runtime/native_type.hpp"
#include "system/app_context.hpp"
#include "system/bit_converter.hpp"
#include "system/diagnostics/debug.hpp"
#include "system/diagnostics/stopwatch.hpp"
#include "system/guid.hpp"
#include "system/io/file-stream.hpp"
#include "system/io/file.hpp"
#include "system/io/memory-stream.hpp"
#include "system/io/path.hpp"
#include "system/io/stream-reader.hpp"
#include "system/io/stream.hpp"
#include "system/io/string-reader.hpp"
#include "system/math.hpp"
#include "system/number.hpp"
#include "system/security/cryptography/sha256.hpp"
#include "system/string_comparer.hpp"
#include "system/text/encoding.hpp"
#include "system/text/regular_expressions/regex.hpp"
#include "system/text/string-builder.hpp"

AnchorComponent::AnchorComponent() : IsSubscribedToWindowResize(), anchorBoundsProvider(), anchorData()
{
}

bool AnchorComponent::get_IsAnchored()
{
return this->anchorData != nullptr;
}

void AnchorComponent::ComponentAdded(::Entity* entity)
{
Component::ComponentAdded(entity);
    if (this->anchorData != nullptr)
    {
this->RefreshSubscriptions();
this->RefreshAnchoring();
    }
}

void AnchorComponent::ComponentRemoved(::Entity* entity)
{
Component::ComponentRemoved(entity);
this->DisableAnchoring();
}

void AnchorComponent::DisableAnchoring()
{
this->DetachFromBoundsProvider();
this->DetachFromWindowResize();
this->anchorData = nullptr;
}

void AnchorComponent::EnableAnchoring(bool left, bool right, bool top, bool bottom)
{
    if (!left && !right && !top && !bottom)
    {
this->DisableAnchoring();
return;    }
    if (Parent == nullptr)
    {
throw new InvalidOperationException("AnchorComponent must be attached before anchoring can be enabled.");
    }
int2 *anchorBounds = this->GetAnchorBounds();
int2 *anchoredSize = this->GetAnchorSize();
::float3 localPosition = Parent->get_LocalPosition();
this->anchorData = ([&]() {
auto __object_0000017A = new ::AnchorData();
__object_0000017A->set_LeftDistance(left ? Nullable<float>(localPosition.X) : Nullable<float>(nullptr));
__object_0000017A->set_RightDistance(right ? Nullable<float>(anchorBounds->X - localPosition.X - anchoredSize->X) : Nullable<float>(nullptr));
__object_0000017A->set_TopDistance(top ? Nullable<float>(localPosition.Y) : Nullable<float>(nullptr));
__object_0000017A->set_BottomDistance(bottom ? Nullable<float>(anchorBounds->Y - localPosition.Y - anchoredSize->Y) : Nullable<float>(nullptr));
return __object_0000017A;
})();
this->RefreshSubscriptions();
this->RefreshAnchoring();
}

std::string AnchorComponent::GetAnchorInfo()
{
    if (!this->get_IsAnchored())
    {
return "Not anchored";    }
std::string info = "Anchored to: ";
List<std::string> *anchors = new List<std::string>();
    if (this->anchorData->get_LeftDistance().HasValue)
    {
anchors->Add(std::string("Left (") + std::to_string(this->anchorData->get_LeftDistance().Value) + std::string("px)"));
    }
    if (this->anchorData->get_RightDistance().HasValue)
    {
anchors->Add(std::string("Right (") + std::to_string(this->anchorData->get_RightDistance().Value) + std::string("px)"));
    }
    if (this->anchorData->get_TopDistance().HasValue)
    {
anchors->Add(std::string("Top (") + std::to_string(this->anchorData->get_TopDistance().Value) + std::string("px)"));
    }
    if (this->anchorData->get_BottomDistance().HasValue)
    {
anchors->Add(std::string("Bottom (") + std::to_string(this->anchorData->get_BottomDistance().Value) + std::string("px)"));
    }
return String::Concat(info, String::JoinArray(", ", anchors->ToArray()));}

void AnchorComponent::RefreshAnchoring()
{
    if (this->anchorData == nullptr || Parent == nullptr)
    {
return;    }
this->RefreshSubscriptions();
int2 *anchorBounds = this->GetAnchorBounds();
int2 *anchorSize = this->GetAnchorSize();
::float3 localPosition = Parent->get_LocalPosition();
    if (this->anchorData->get_LeftDistance().HasValue)
    {
localPosition.X = this->anchorData->get_LeftDistance().Value;
    }
else     if (this->anchorData->get_RightDistance().HasValue)
    {
localPosition.X = anchorBounds->X - this->anchorData->get_RightDistance().Value - anchorSize->X;
    }
    if (this->anchorData->get_TopDistance().HasValue)
    {
localPosition.Y = this->anchorData->get_TopDistance().Value;
    }
else     if (this->anchorData->get_BottomDistance().HasValue)
    {
localPosition.Y = anchorBounds->Y - this->anchorData->get_BottomDistance().Value - anchorSize->Y;
    }
Parent->set_LocalPosition(localPosition);
}

void AnchorComponent::SetAnchorDistances(Nullable<float> left, Nullable<float> right, Nullable<float> top, Nullable<float> bottom)
{
    if (this->anchorData == nullptr)
    {
this->anchorData = new ::AnchorData();
    }
this->anchorData->set_LeftDistance(left);
this->anchorData->set_RightDistance(right);
this->anchorData->set_TopDistance(top);
this->anchorData->set_BottomDistance(bottom);
    if (!left.HasValue && !right.HasValue && !top.HasValue && !bottom.HasValue)
    {
this->DisableAnchoring();
return;    }
this->RefreshSubscriptions();
this->RefreshAnchoring();
}

::Entity* AnchorComponent::get_Parent()
{
return this->Component::get_Parent();
}

void AnchorComponent::set_Parent(::Entity* value)
{
this->Component::set_Parent(value);
}

void AnchorComponent::AttachToWindowResize()
{
    if (this->IsSubscribedToWindowResize)
    {
return;    }
Core::get_Instance()->get_RenderManager3D()->WindowResized += &AnchorComponent::HandleWindowResized;
this->IsSubscribedToWindowResize = true;
}

void AnchorComponent::DetachFromBoundsProvider()
{
    if (this->anchorBoundsProvider != nullptr)
    {
this->anchorBoundsProvider->AnchorBoundsChanged -= &AnchorComponent::HandleAnchorBoundsChanged;
this->anchorBoundsProvider = nullptr;
    }
}

void AnchorComponent::DetachFromWindowResize()
{
    if (!this->IsSubscribedToWindowResize)
    {
return;    }
Core::get_Instance()->get_RenderManager3D()->WindowResized -= &AnchorComponent::HandleWindowResized;
this->IsSubscribedToWindowResize = false;
}

int32_t AnchorComponent::GetAnchorArea(int2* size)
{
    if (size->X < 0 || size->Y < 0)
    {
return -1;    }
return size->X * size->Y;}

int2* AnchorComponent::GetAnchorBounds()
{
    if (this->anchorBoundsProvider != nullptr)
    {
return this->anchorBoundsProvider->get_AnchorBounds();    }
return Core::get_Instance()->get_RenderManager3D()->get_MainWindowSize();}

int2* AnchorComponent::GetAnchorSize()
{
    if (Parent == nullptr)
    {
return new int2(0, 0);    }
::IAnchorSizeProvider *bestProvider = nullptr;
int32_t bestArea = -1;
    IAnchorSizeProvider* parentProvider = he_cpp_try_cast<IAnchorSizeProvider>(Parent);
    if (parentProvider != nullptr)
    {
bestProvider = parentProvider;
bestArea = this->GetAnchorArea(parentProvider->get_AnchorSize());
    }
for (int32_t i = 0; i < Parent->get_Components()->Count(); i++) {
    IAnchorSizeProvider* sizeProvider = he_cpp_try_cast<IAnchorSizeProvider>((*Parent->get_Components())[i]);
    if (sizeProvider != nullptr)
    {
const int32_t area = this->GetAnchorArea(sizeProvider->get_AnchorSize());
    if (area > bestArea)
    {
bestProvider = sizeProvider;
bestArea = area;
    }
    }
}
    if (bestProvider == nullptr)
    {
return new int2(0, 0);    }
return bestProvider->get_AnchorSize();}

void AnchorComponent::HandleAnchorBoundsChanged()
{
this->RefreshAnchoring();
}

void AnchorComponent::HandleWindowResized(intptr_t handle, int32_t newWidth, int32_t newHeight)
{
this->RefreshAnchoring();
}

void AnchorComponent::RefreshSubscriptions()
{
::IAnchorBoundsProvider *newProvider = this->ResolveAnchorBoundsProvider();
    if (!(this->anchorBoundsProvider == newProvider))
    {
this->DetachFromBoundsProvider();
this->anchorBoundsProvider = newProvider;
    if (this->anchorBoundsProvider != nullptr)
    {
this->anchorBoundsProvider->AnchorBoundsChanged += &AnchorComponent::HandleAnchorBoundsChanged;
    }
    }
    if (this->anchorBoundsProvider == nullptr)
    {
this->AttachToWindowResize();
    }
else {
this->DetachFromWindowResize();
}
}

::IAnchorBoundsProvider* AnchorComponent::ResolveAnchorBoundsProvider()
{
::Entity *current = Parent;
while (current != nullptr) {
    IAnchorBoundsProvider* provider = he_cpp_try_cast<IAnchorBoundsProvider>(current);
    if (provider != nullptr)
    {
return provider;    }
current = current->get_Parent();
}
return nullptr;}

